Task3/Task3_Jordan_Petrov/src/python/car_wash/pool.py

```diff
@@ -0,0 +1,121 @@
+"""
+Generic db pooling code for two types of db users:
+ - read and write,
+ - read only
+You can use them as context manager
+You can run 'python3 pool.py -h' see how to parse parameters from console
+"""
+import argparse
+import logging
+
+import psycopg2
+
+from config import READ_ONLY_USER, READ_WRITE_USER
+from utils import SingletonMeta
+
+logging.basicConfig(
+    filename='../../../pool.log',
+    format='%(asctime)s - %(message)s',
+    level=logging.INFO
+)
+
+
+class BaseConnectionPool(metaclass=SingletonMeta):
+    """
+    Base class for the read only and read write classes
+    """
+
+    def __init__(self, conns: int, *args, **kwargs):
+        """
+        Class constructor
+        param conns: number of open connections
+        param args: passed to psycopg2.connect
+        param kwargs: passed to psycopg2.connect
+        """
+        self._kwargs = kwargs
+        self._args = args
+        self.conns = int(conns)
+        self._pool: list[psycopg2.connect] = []
+
+        for _ in range(self.conns):
+            self._connect()
+        logging.info(f'Connection to db established, added needed connections to the pool.')
+
+    def __enter__(self):
+        """
+        Allows to use the connection pool as a context manager.
+        Use cursor to execute queries.
+        """
+        self.connection = self._get_connection()
+        self.cursor = self.connection.cursor()
+        return self.cursor
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        """
+        Close cursor, commit connection and returns the connection to the pool.
+        """
+        self.connection.commit()
+        self.cursor.close()
+        self._put_connection(self.connection)
+
+    def _connect(self) -> psycopg2.connect:
+        """Create a new connection"""
+        try:
+            conn = psycopg2.connect(*self._args, **self._kwargs)
+        except psycopg2.Error:
+            logging.error('Invalid connection parameters.')
+            raise psycopg2.Error('Invalid connection parameters.')
+        self._pool.append(conn)
+        return conn
+
+    def _get_connection(self) -> psycopg2.connect:
```

Comment:
I propose to modify this method:
- add while loop to wait on free connection
- add `with lock` to get a connection from pool
-  open new connection  when it will be required
- add the check when we created max number of connection and we must wait